src/ck-core/ckrdma.h

```diff
@@ -424,4 +469,97 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr);
 void CkRdmaZCPupCustomHandler(void *ack);
 
 void _ncpyAckHandler(ncpyHandlerMsg *msg);
+
+void setNcpyEmInfo(NcpyEmInfo *ref, envelope *env, int &numops, void *forwardMsg, ncpyEmApiMode emMode);
+
+// Struct is used to store posted buffer information
+// Information passed in a CkPostBuffer call is used to construct CkPostedBuffer
+struct CkPostedBuffer {
+  void *buffer;
+  int bufferSize;
+};
+
+// Struct passed in a ZC Post Entry Method to allow receiver side to post
+// This structure variable is passed in a CkMatchBuffer call in the Post EM
+struct CkNcpyBufferPost {
+  // regMode
+  unsigned short int regMode;
+
+  // deregMode
+  unsigned short int deregMode;
+
+  // index within message
+  int index;
+
+  bool postAsync;
+
+  size_t tag;
+
+  // NcpyEmInfo
+  NcpyEmInfo *ncpyEmInfo;
+
+  void *srcBuffer;
+
+  size_t srcSize;
+
+  int opIndex;
+
+  CmiUInt8 arrayIndex;
+};
+
+void handleArrayMsgOnChildPostCompletionForRecvBcast(envelope *env, NcpyEmInfo *ncpyEmInfo);
+
+void handleGroupMsgOnChildPostCompletionForRecvBcast(envelope *env, NcpyEmInfo *ncpyEmInfo);
+
+void handleNGMsgOnChildPostCompletionForRecvBcast(envelope *env, NcpyEmInfo *ncpyEmInfo);
+
+void CkMatchBuffer(CkNcpyBufferPost *post, int index, int tag);
+
+void CkMatchNodeBuffer(CkNcpyBufferPost *post, int index, int tag);
+
+void updatePeerCounter(NcpyEmInfo *ncpyEmInfo);
+
+void updateTagArray(envelope *env, int localElems);
+
+void setPosted(std::vector<int> *tagArray, envelope *env, CmiUInt8 elemIndex, int numops, int opIndex);
+
+bool isUnposted(std::vector<int> *tagArray, envelope *env, CmiUInt8 elemIndex, int numops);
+
+int extractStoredBuffer(std::vector<int> *tagArray, envelope *env, CmiUInt8 elemIndex, int numops, int opIndex, void *&ptr);
+
+int CkPerformRget(CkNcpyBufferPost &post, void *destBuffer, int destSize);
+
+void setPostStruct(CkNcpyBufferPost *ncpyPost, int index, CkNcpyBuffer &buffObj, CmiUInt8 elemIndex);
+
+void initPostStruct(CkNcpyBufferPost *ncpyPost, int index);
+
+void CkPostBufferInternal(void *destBuffer, size_t destSize, int tag);
+void CkPostNodeBufferInternal(void *destBuffer, size_t destSize, int tag);
+
+template <typename T>
+static inline constexpr size_t safe_sizeof(T * ptr)
+{
+    return sizeof(T);
+}
+
+template <>
+inline constexpr size_t safe_sizeof(void * ptr)
+{
+    return 1;
+}
```

Comment:
Since this is in a .h file, it would be good to put these in `namespace details` to avoid potential conflicts since this is a fairly generic name. There is an example of this in `src/util/pup.h`.

Reply:
That's a good point, added in the namespace. There are a lot of other global functions in the ckrdma implementation, I'll add another patch to wrap up those in a namespace.